File: containers/controller/controller.py
import json
import logging
import socket
import threading
from datetime import datetime
from redisController import redisController
from networking import *
import time
import queue
import uuid

DB_HOST = "redis"
DB_PORT = 6379
CONTROLLER_PORT = 8080

logging.basicConfig(level=logging.DEBUG,
                    format='%(asctime)s.%(msecs)d:LINE %(lineno)d:TID %(thread)d:%(levelname)s - %(message)s', datefmt='%d-%b-%y %H:%M:%S')

modes = ["serial", "parallel"]
# Thread independent hash of container ids and a socket, so we can send stuff.
# I am using thread locks instead of another thread + a queue, much nicer :-)
connections = {}


def isSupportedMode(mode):
    return mode in modes


def handleAction(action, mode, read_queue, q):
    # check redis and make sure its actually an action that exists...if it doesn't a lot we'll need to look at queues or something...

    redis_action = rc.queryAction(action)
    if redis_action is None:
        logging.warn("Action: " + action +
                     " does not exist (yet) in redis...cannot call action. Waiting on queue instead...")
        # block on this queue until the action we are looking for is registered
        while q.get() != action:
            logging.debug(
                "We didn't find it in redis, but something came in the queue, but it was not the action we wanted...")
        # now check the db, if its not there then something is internally wrong
        logging.debug("We got the action we wanted to see...")
        redis_action = rc.queryAction(action)
        if redis_action is None:
            logging.error(
                "WTF, how can it not be there? We found the action in the 'just registered' queue")

    # Validate the fields are there that we care about...mainly the container_id
    try:
        container_id = redis_action['container_id']
        logging.debug("Found action: " + action)
        logging.debug("Container who owns action is: " + container_id)
        # check the mode they passed into this function: (parallel vs serial to start)
        if not isSupportedMode(mode):
            logging.error(
                "The mode passed in is not a supported mode!!! Treating as serial...")
        # Do something for these modes, default to serial

        # No matter what, we send something to the proper container
        message = {
            'type': "trigger-action",
            'timestamp': time.time(),
            'data': {
                'name': action
            }
        }
        # Determine which socket we need to send this to and lock the sock!
        try:
            logging.debug(
                "Dumping all things in the 'connections' structure...")
            global connections
            logging.debug(connections)
            connection = connections[container_id]
            write_lock = connection['write_lock']
            socket = connection['socket']
            read_queue = connection['read_queue']
            with write_lock:
                # No matter the mode, we want to send something!
                socket.sendall(packetize(json.dumps(message)))
            if mode == "parallel":
                # Don't block per action, we don't care what the result is (yet)
                # The responses will go into the read_queue though, for later
                return
            else:  # if mode == "serial" or something else...
                # wait for a reply from the socket that is a result from this request (this should block)
                response = read_queue.get()
                return response
        except KeyError as e:
            logging.debug(e)
            logging.error(
                "Not sure how, but this connection does not have a connection in the connection hash!")

    except KeyError as e:
        logging.debug(
            "Could not find a container_id in the redis action of " + action)


def handleEvent(obj, rc, read_queue, q):
    # Internal error if we somehow don't go through the if or else
    response = {
        'type': "emit-event-error",
        'timestamp': time.time(),
        'data': {
                'message': "Internal emit-event error",
                'status': 506
        }
    }
    # Grab the timestamp in the packet
    timestamp = obj['timestamp']
    # Go the event being handled
    event = obj['data']['event']
    # Pull out the valuable attributes from this layer
    event_name = event['name']
    # There may or may not be a payload per event, we should pass it if it exists..., for now don't worry
    # event_payload = event['payload']

    container_id = obj['container_id']
    # Query out the actions that take place because of this event
    redis_event = rc.queryEvent(event_name)
    logging.debug(
        "Checking for an existing event returned: " + str(json.dumps(redis_event)))
    redis_parsed = {}

    # Verify the container who emitted this event is the one who registered it
    if redis_event is None:
        response['data']['message'] = "Cannot emit an event which does not exist"
        response['data']['status'] = 508
    elif container_id != redis_event['container_id']:
        response['data']['message'] = "container_id of request did not match container_id of registered event"
        response['data']['status'] = 507
    else:
        # Handle the ignore
        try:
            ignore = redis_event['ignore']
            if isinstance(ignore, str):
                rc.ignoreEvent(ignore)
            elif isinstance(ignore, list):
                for i in ignore:
                    rc.ignoreEvent(ignore)
            else:
                logging.error("Could not handle ignore for event: " +
                              event_name + ". Not string or list!!!")
        except KeyError as e:
            logging.debug(
                "No ignore found while parsing event: " + event_name)
        try:
            listen = redis_event['listen']
            if isinstance(listen, str):
                rc.listenEvent(listen)
            elif isinstance(listen, list):
                for l in listen:
                    rc.listenEvent(l)
        except KeyError as e:
            logging.debug(
                "No listen found while parsing event: " + event_name)
        try:
            # We use this later when doing a serial execution
            brk = redis_event['break']
        except KeyError as e:
            logging.debug(
                "No break found while parsing event: " + event_name)
        # You need either a serial or parallel. No support for both yet
        try:
            serial = redis_event['serial']
            # For every serial action
            for action in serial:
                logging.debug(
                    "Serial Action being called: " + str(action))
                # Call that action and block until we get a response (block happens in the function)
                handleAction(action, "serial", read_queue, q)
        except KeyError as e:
            logging.debug(
                "No serial found while parsing event: " + event_name)
            try:
                parallel = redis_event['parallel']
                for action in parallel:
                    logging.debug(
                        "Parallel Action being called: " + str(action))
                    # Call that action but don't block waiting for a response, instead, collect
                    # them outside of here (by draining the queue later)
                    # TODO this needs to become a thread per action so we don't block on missing actions
                    # like we would on serial
                    handleAction(action, "parallel", read_queue, q)
                # For now, drain the queue here
                logging.debug("Draining the parallel results:")
                while not read_queue.empty():
                    logging.debug(read_queue.get())
            except KeyError as e:
                logging.debug(
                    "No parallel found while parsing event: " + event_name)
        response = {
            'type': "emit-event-response",
            'timestamp': time.time(),
            'data': {
                'message': "OK",
                'status': 0
            }
        }
    return response


def handle_container_message(client_socket, client_address, container_object, rc, connection, events):
    # This function makes the appropriate calls to other functions based on the context of the packet

    # Make a generic response object in case something goes wrong where we don't hit anything else
    response = {
        'type': "generic-error-response",
        'timestamp': time.time(),
        'data': {
            'message': "Internal controller handler error",
            'status': 503
        }
    }

    if container_object['type'] == "register-container":
        response = rc.registerContainer(container_object)
        if response['data']['status'] == 0:
            # If we got a good register container response, add it to the list of lock n socks...
            # We'll likely be communicating with it soon and we need to lock the sock!
            try:
                container_id = container_object['data']['container']['id']
                global connections
                connections[container_id] = connection
            except KeyError as e:
                logging.error(e)
                logging.error(
                    "KeyError when trying to add connection to the global connection list")
    elif container_object['type'] == "register-event":
        response = rc.registerEvent(container_object)
    elif container_object['type'] == "register-action":
        response = rc.registerAction(container_object, events)
    elif container_object['type'] == "emit-event":
        # Events may block because they trigger a serial process, need to spawn a thread
        # There also needs to be an event id to prevent overlaps from occuring. This will go in the action packets to keep track what event this is for
        event_id = str(uuid.uuid4())
        # add our queue to the events hash with the key being the event_id
        events[event_id] = {
            'name': container_object['data']['event'],
            'read_queue': queue.Queue(),
            'action_queue': queue.Queue()
        }
        # Spawn that thread
        t = threading.Thread(target=handleEvent, args=(
            container_object, rc, events[event_id]['read_queue'], events[event_id]['action_queue']))
        t.start()
        response = {
            'type': "emit-event-response",
            'timestamp': time.time(),
            'data': {
                'message': "OK",
                'status': 0
            }
        }
        #handleEvent(container_object, rc, read_queue)
    elif container_object['type'] == "trigger-action-response":
        # We know that an emit-event thread is listening for this response. We'll put this in their queue so it unblocks
        # Figure out who this owns to based on the data
        try:
            # Read what event_id this is
            event_id = container_object['event_id']
            # Pull the queue for this event id
            q = events[event_id]
            # Send what we got to the proper queue, which should unblock that event process
            q.put(container_object)
            # No response needed since this is a response
            return
        except KeyError as e:
            logging.error(
                "We're missing the event id, can't determine which queue is blocked / waiting for this")
    else:
        response = {
            'type': container_object['type'] + "-error",
            'timestamp': time.time(),
            'data': {
                'message': "unrecognized type: " + container_object['type'],
                'status': 1
            }
        }
    return response


def serve_container(client_socket, client_address, rc):
    try:
        # We have a client and they've connected to us
        connected = True
        # Create locks for this container
        connection = {
            'socket': client_socket,
            'write_lock': threading.Lock(),
            'read_queue': queue.Queue()
        }
        # Hash of current events and their queues because we could be handling
        # multiple events at similar times, ones blocked, one's not, yikes! This is insane
        events = {}

        # While they are here...
        while connected:
            # Block until we get a packet from them
            logging.debug("Blocking for " + client_address +
                          " waiting to receive a packet...")
            # This is the only place we read a packet, so no lock is needed.
            raw_packet, err = receive_packet(client_socket)
            # See if they left
            if(err):
                logging.warning("Container at " + client_address + "has left.")
                connected = False
                break
            # If they didn't, we must have a message, and we should load it into a JSON object
            logging.debug("Parsing packet from " + client_address + "...")
            container_object = json.loads(depacketize(raw_packet))
            logging.debug("Packet from " + client_address + " is...")
            logging.debug(container_object)
            logging.debug("Handling JSON for " + client_address)
            response = handle_container_message(
                client_socket, client_address, container_object, rc, connection, events)
            # Now we can send the response back, if there is something
            if response is not None:
                with connection['write_lock']:
                    client_socket.sendall(packetize(json.dumps(response)))
    except socket.timeout:
        logging.debug("Lost connection to " + client_address)
        client_socket.close()

# Main


# Create the server socket
logging.debug("Creating the server socket on port: " + str(CONTROLLER_PORT))
server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
server_socket.bind(("0.0.0.0", CONTROLLER_PORT))
logging.info("Controller server up and listening on port: " +
             str(CONTROLLER_PORT) + ".")

# Connect to the database
rc = redisController(DB_HOST, DB_PORT)

# Path to the config yaml
rc.setConfig("config.yml")

# Listen for other containers to connect
server_socket.listen()
while True:
    logging.debug("Waiting for containers on the main thread...")
    client_socket, client_address = server_socket.accept()
    client_address = str(client_address)
    logging.info("Received a connection from: " + client_address + ".")
    logging.debug(
        "Spawning a new thread to handle communication with " + client_address + ".")
    t = threading.Thread(target=serve_container,
                         args=(client_socket, client_address, rc))
    t.start()

File: containers/controller/controller.go
package main

import (
	common "controller/common"
	parser "controller/parser"
	"encoding/json"
	"errors"
	"log"
	"net"
	"net/http"
	_ "net/http/pprof"
	"os"
	"strconv"
	"time"

	"github.com/jinzhu/copier"
	"github.com/op/go-logging"
)

const (
	//REGISTERCONTAINER -
	REGISTERCONTAINER = "register-container"
	//REGISTERCONTAINERRESPONSE -
	REGISTERCONTAINERRESPONSE = "register-container-response"
	//REGISTEREVENT -
	REGISTEREVENT = "register-event"
	//REGISTEREVENTRESPONSE -
	REGISTEREVENTRESPONSE = "register-event-response"
	//REGISTERACTION -
	REGISTERACTION = "register-action"
	//REGISTERACTIONRESPONSE -
	REGISTERACTIONRESPONSE = "register-action-response"
	//EMITEVENT -
	EMITEVENT = "emit-event"
	//DISPATCHEDEVENT -
	DISPATCHEDEVENT = "dispatched-event"
	//EMITEVENTRESPONSE -
	EMITEVENTRESPONSE = "emit-event-response"
	//TRIGGERACTION -
	TRIGGERACTION = "trigger-action"
	//TRIGGERACTIONRESPONSE -
	TRIGGERACTIONRESPONSE = "trigger-action-response"
	//OK -
	OK = 200
	//ERROR -
	ERROR = 503
	//ONLINE -
	ONLINE = "online"
	//OFFLINE -
	OFFLINE = "offline"
)

//Conn - local conn that handles output with a channel
type conn struct {
	name string
	conn net.Conn
	out  chan []byte
	in   chan []byte
}

func (cio *conn) HandleConnOut(c net.Conn) {
	for {
		b := <-cio.out
		b = append(b, '\n')
		c.Write(b)
	}
}

//buildResponse - takes a message that came in and turn it into a response
func (cio *conn) buildResponse(msg *common.Message, code int, dispatched bool) {
	var m common.Message
	copier.Copy(&m, msg)
	m.ResponseCode = code
	m.Timestamp = time.Now().Unix()
	//set the response code based on the message code
	switch msg.Type {
	case REGISTERCONTAINER:
		m.Type = REGISTERCONTAINERRESPONSE
	case REGISTEREVENT:
		m.Type = REGISTEREVENTRESPONSE
	case REGISTERACTION:
		m.Type = REGISTERACTIONRESPONSE
	case EMITEVENT:
		if dispatched {
			m.Type = DISPATCHEDEVENT
		} else {
			m.Type = EMITEVENTRESPONSE
		}
	default:
		panic("buildResponse does not understand how to respond to " + m.Type)
	}
	b, err := json.Marshal(&m)
	if err != nil {
		panic(err) //was unable to build a response
	}
	cio.out <- b
}

//Controller - controls the whole scope of containers
type Controller struct {
	logger      *logging.Logger
	config      *parser.Config
	connections map[string]conn
}

/*
Example Messages:
	Register a Container:
		Type=register-container
		Timestamp=15707945
		ContainerName=gpio
		Name=gpio
		ResponseCode=nil
		Properties=nil
	Response:
		Type=register-container-response
		Timestamp=15707946
		ContainerName=gpio
		Name=gpio
		ResponseCode=200
		Properties=nil
	Emitting an Event:
		Type=emit-event
		Timestamp=15707947
		ContainerName=obdii
		Name=speed_changed
		ResponseCode=nil
		Properties={
			speed=50
			unit=mph
		}
	Trigger an Action:
		Type=trigger-action
		Timestamp=15707948
		ContainerName=espeak
		Name=speak
		ResponseCode = nil
		Properties={
			message=You're going too fast
		}
	Get an action response:
		Type=trigger-action-response
		Timestamp=15707949
		ContainerName=espeak
		Name=speak
		ResponseCode=200
		Properties = nil
*/

func (c *Controller) registerContainer(msg *common.Message, cio conn) {
	err := c.config.RegisterContainer(msg)
	code := OK
	if err != nil {
		log.Println(err)
		code = ERROR
	} else {
		//Now that they've registered, take the name they registered with and add it to the list of controller connections in the map
		c.connections[msg.Name] = cio
	}
	//Send a response
	cio.buildResponse(msg, code, false)
}

func (c *Controller) registerAction(msg *common.Message, cio conn) {
	err := c.config.RegisterAction(msg)
	code := OK
	if err != nil {
		log.Println(err)
		code = ERROR
	}
	//Send a response
	cio.buildResponse(msg, code, false)
}

func (c *Controller) registerEvent(msg *common.Message, cio conn) {
	err := c.config.RegisterEvent(msg)
	code := OK
	if err != nil {
		log.Println(err)
		code = ERROR
	}
	//Send a response
	cio.buildResponse(msg, code, false)
}

//triggerSerialAction reaches out to a container and tells it to do something serially
func (c *Controller) triggerSerialAction(act *parser.Action, id string) *common.Message {
	//TODO - actually trigger actions by reading the event config
	var err error
	var msg common.Message
	if err != nil {
		panic(err)
	}

	//Send a trigger action to the container that owns this action
	msg.Type = TRIGGERACTION
	msg.Timestamp = time.Now().Unix()
	msg.ContainerName = act.Container
	msg.Properties = act.Parameters
	msg.Name = act.Name
	msg.ID = id

	b, err := json.Marshal(&msg)
	if err != nil {
		panic(err) //was unable to build a message
	}
	c.connections[act.Container].out <- b
	log.Println("Sent action to " + act.Container)

	//Wait for the response here
	retMsg := c.findYourResponse(act)
	return &retMsg
}

func (c *Controller) findYourResponse(act *parser.Action) common.Message {
	var retmsg common.Message
	//Loop through things in the input queue in case we get a different event's actionresponse (just to be safe)
	for {
		bretmsg := <-c.connections[act.Container].in
		// log.Println(bretmsg)
		err := json.Unmarshal(bretmsg, &retmsg)
		if err != nil {
			panic(err)
		}
		if retmsg.ContainerName != act.Container || retmsg.Name != act.Name {
			//We pulled something from the channel that isn't for here, put it back on the queue
			c.connections[act.Container].in <- bretmsg
		} else {
			break
		}
	}
	retmsg.ContainerName = act.Container
	retmsg.Name = act.Name
	retmsg.Type = TRIGGERACTIONRESPONSE
	return retmsg
}

//triggerParallelAction reaches out to a container and tells it to do something in parallel
func (c *Controller) triggerParallelAction(act *parser.Action, id string, ret chan common.Message) {
	//TODO - actually trigger actions by reading the event config
	var err error
	var msg common.Message
	if err != nil {
		panic(err)
	}

	//Send a trigger action to the container that owns this action
	msg.Type = TRIGGERACTION
	msg.Timestamp = time.Now().Unix()
	msg.ContainerName = act.Container
	msg.Properties = act.Parameters
	msg.ID = id
	msg.Name = act.Name

	b, err := json.Marshal(&msg)
	if err != nil {
		panic(err) //was unable to build a message
	}

	//Make sure the container & action we want to send to is online

	c.connections[act.Container].out <- b
	log.Println("Sent action to " + act.Container)

	retmsg := c.findYourResponse(act)
	ret <- retmsg
}

//handleEvent is the part of the controller responsible for goroutines that do all kinds of processes
func (c *Controller) handleEvent(msg *common.Message, cio conn) {
	//The message we were given told use to emit an event
	log.Println("We're starting event: ", msg.Name)

	//Make a unique id for this event, so everyone can keep track
	uuid := common.GenUUID()
	//Send them a dispatched message so they know we're handling their message
	msg.ID = uuid
	cio.buildResponse(msg, OK, true)

	//Pull the information about who what containers we should contact
	event, err := c.config.GetEvent(msg.Name)
	if err != nil {
		log.Println(err)
	} else if event.State == "offline" {
		log.Println(errors.New("Event " + event.EventName + " emitted before being registered"))
	} else {
		//We have a valid event we can analyze
		blocks := *event.Blocks
		run := true  //See if the conditionals warrant a run of this event
		good := true //Set to false if anything bad happens when dispatching an event
		for _, b := range blocks {
			switch b.Type {
			case "when", "and", "or":
				//Conditionals
				for i, children := range b.Children {
					switch cond := children.(type) {
					//Look for the type of each conditional
					case parser.Condition:
						//If we should still run it this condition should be true
						if i == 0 {
							if b.Type != "when" {
								//Panic, when should be the first conditional we see
								good = false
								log.Println("when should be the first conditional for " + event.EventName)
							} else {
								//This is our first conditional, just set run to the evaluation
								run = c.config.EvaluateCondition(cond)
							}
						} else {
							//If the type is anded, we should and it with run
							if b.Type == "and" {
								run = run && c.config.EvaluateCondition(cond)
							} else if b.Type == "or" {
								run = run || c.config.EvaluateCondition(cond)
							} else {
								good = false
								log.Println("Unrecognized conditional: " + b.Type)
							}
						}
					default:
						good = false
						log.Println("Saw a recognized type: " + b.Type + ", but did not .(type) to a conditional")
					}
				}
			case "serial", "parallel":
				//No-op, don't error but don't process until after we get an idea of the run status
			default:
				good = false
				log.Println("Unknown time when attempting to emit an event: " + b.Type)
			}

		}
		//Only loop through instruction blocks
		if run {
			for _, b := range blocks {
				var channels []chan common.Message
				switch b.Type {
				case "when", "and", "or":
					//no-op
				case "serial", "parallel":
					//Blocks of instructions, we'll only run this when we get out of the block loop and check the value of run
					for _, children := range b.Children {
						switch act := children.(type) {
						case *parser.Action:
							log.Println("About to execute '" + act.Name + "' in " + b.Type)
							if b.Type == "serial" {
								ret := c.triggerSerialAction(act, uuid)
								if ret.ResponseCode != OK {
									log.Println("Got a bad return code from serial response: " + ret.Name + ", " + strconv.Itoa(ret.ResponseCode))
									good = false
								}
								log.Println("Serial job finished: " + ret.Name)
							} else if b.Type == "parallel" {
								channel := make(chan common.Message)
								//Keep track of all the channels we've made that will be returning something
								channels = append(channels, channel)
								go c.triggerParallelAction(act, uuid, channel)
							} else {
								good = false
								log.Println("Found serial/parallel earlier but now can't find it, this must be a coder error")
							}
						default:
							good = false
							log.Println("Expected action after entering an instruction block")
						}
					}
					//After we loop through and we are parallel, we want to block until all the parallel things are done
					if b.Type == "parallel" {
						//Handle the messages from the goroutines
						for _, channel := range channels {
							ret := <-channel
							if ret.ResponseCode == OK {
								//This is good
								log.Println("Parallel job finished: " + ret.Name)
							} else {
								log.Println("Got a bad return code from parallel response: " + ret.Name + ", " + strconv.Itoa(ret.ResponseCode))
								good = false
							}
						}
					}
				default:
					panic("Unknown type when running through instruction blocks after a successful run condition")
				}
			}
			//We ran, and everything we dispatched came back, ready to respond to the caller
			if good {
				cio.buildResponse(msg, OK, false)
			} else {
				cio.buildResponse(msg, ERROR, false)
			}
		} else {
			//We did not run, but we didn't run by the conditional, not by error
			if good {
				cio.buildResponse(msg, OK, false)
			} else {
				cio.buildResponse(msg, ERROR, false)
			}
		}
		log.Println("We're finishing event: ", msg.Name)
	}
}

//handleActionResponse - sends action response to proper connection's pseduo-input channel
func (c *Controller) handleActionResponse(msg *common.Message, cio conn) {
	b, err := json.Marshal(msg)
	if err != nil {
		c.logger.Error(err)
	} else {
		c.connections[msg.ContainerName].in <- b
	}
}

func (c *Controller) handleConnection(conn net.Conn, cio conn) {
	defer conn.Close()
	c.logger.Infof("Handling %s\n", conn.RemoteAddr().String())
	//Spin up a goroutine to write out to this connection (essentially an output queue)
	go cio.HandleConnOut(conn)
	//Handle input here directly from the socket
	for {
		// read directly from the socket, expecting each json message to be newline separated
		d := json.NewDecoder(conn)
		var msg common.Message
		err := d.Decode(&msg)
		if err != nil {
			c.logger.Error(err)
			break
		}
		// spew.Dump(msg)
		//Read the type and send it to the proper function for further processing
		switch msg.Type {
		case REGISTERCONTAINER:
			go c.registerContainer(&msg, cio)
		case REGISTERACTION:
			go c.registerAction(&msg, cio)
		case REGISTEREVENT:
			go c.registerEvent(&msg, cio)
		case EMITEVENT:
			go c.handleEvent(&msg, cio)
		case TRIGGERACTIONRESPONSE:
			go c.handleActionResponse(&msg, cio)
		default:
			c.logger.Error("Unknown Type:", msg.Type)
			break
		}
		if err != nil {
			panic(err)
		}
	}
}

func (c *Controller) setupLogger() {
	c.logger = logging.MustGetLogger("main")
	formatter := logging.MustStringFormatter(
		`%{color}%{time:15:04:05.00} %{shortfunc} ▶ %{level:.4s} %{id:03x}%{color:reset} %{message}`)

	backend := logging.NewLogBackend(os.Stdout, "", 0)
	backendFormatter := logging.NewBackendFormatter(backend, formatter)
	logging.SetBackend(backendFormatter)
	logging.SetLevel(logging.DEBUG, "main")
}
func (c *Controller) readConfig() {
	c.config = c.config.NewConfig()
	config, err := c.config.Parse("/config.yml")
	if err != nil {
		panic(err)
	}
	c.config = config
}

func (c *Controller) setupListener(port int) net.Listener {
	PORT := ":" + strconv.Itoa(port)
	l, err := net.Listen("tcp4", PORT)
	if err != nil {
		panic(err)
	}
	return l
}

//Start - starts a controller
func (c *Controller) Start(port int) {
	c.setupLogger()
	c.readConfig()
	l := c.setupListener(port)
	defer l.Close()
	c.logger.Info("Controller is up. Listening for clients on port", port)
	for {
		con, err := l.Accept()
		if err != nil {
			c.logger.Error(err)
		}
		var cio conn
		//create an input and output channel for this connection
		in := make(chan []byte)
		out := make(chan []byte)
		//Assign them
		cio.out = out
		cio.in = in
		// For every connection that comes in, start a goroutine to handle their inputs
		go c.handleConnection(con, cio)
	}
}

func main() {
	c := Controller{}
	c.connections = make(map[string]conn)
	go func() {
		log.Println(http.ListenAndServe("localhost:6060", nil))
	}()
	c.Start(8080)

}
